binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+# Binary data formats {#sec-binary-data}
+
+```{r setup}
+#| results: "asis"
+#| echo: false
+source("_common.R")
+status("drafting")
+```
+
+## Introduction
+
+### Prerequisites
+
+This chapter focuses on large data sets, and binary file formats that are well suited to working with them.
+We'll use a data set of item checkouts from Seattle public libraries, available online at [data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6](https://data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6).
+The data is distributed as a single 9GB csv file, so it may take some time to download.
```

Comment:
Should we make this available in an S3 bucket (or similar)? And maybe suggest some R code to download it?

Maybe the first section in the chapter should be "getting big datasets", where you could show how to download a file with code and then maybe move in the contents of "Data in the cloud"

Reply:
Agreed with this — adding details of where | how to get this would be fantastic. And if we presume that someone will already have (or be willing to install) arrow in this chapter, we could utilize [`copy_files()`](https://arrow.apache.org/docs/r/reference/copy_files.html) which IMHO is a pretty nice way to fetch files from S3

Reply:
@jonkeane does that have a progress bar? and some ability to resume if the connection is interrupted?

Reply:
IMO, it should do both — though AFAIK it doesn't currently.

We have an issue for a progress bar: https://issues.apache.org/jira/browse/ARROW-14611 (though there was some hesitancy there that it might be out of scope for the Arrow C++ internals)

So maybe ultimately we should use a different method of fetching these that has better UX than `copy_files` currently does?

Reply:
Yeah, that would be nice, but I don't know of anything off the rack 😞

Reply:
+1 for starting the chapter with getting some large (or subset of large) data. If the `copy_files` UX is too minimal (however it is my go-to for S3/GCS at this time), there is the option of letting the reader manually download the data _or_ point `open_dataset()` to the example data hosted in an S3 bucket. There are trade-offs -- the latter can be _slow_ but also means not downloading anything.

Reply:
Given the concerns expressed about the intended scope of the arrow FileSystem interface, and the intended audience of R4DS, it might be best to avoid using `copy_files()` in this chapter; and relatedly, I'm now leaning towards replacing the section talking about Arrow cloud support with a more generic sentence that mentions that this support exists and then links out to the relevant vignette in the Arrow docs.

In the meantime, I'm wondering if the best approach is to host a copy of the csv file in an s3 bucket and then fall back on `download.file()` as the example of how to download?